segmentation/utils/dimensionality_reduction.py

```diff
@@ -0,0 +1,74 @@
+import pandas as pd
+import numpy as np
+from sklearn.preprocessing import StandardScaler
+from sklearn.decomposition import PCA
+from sklearn.manifold import TSNE
+import matplotlib.pyplot as plt
+import seaborn as sns
+import umap
+
+
+def visualize_dimensionality_reduction(cell_data, columns, category, color_map="Spectral", algorithm="UMAP",
+                                       save_dir=None):
+    """Plots the dimensionality reduction of specified population columns
+                Args:
+                    cell_data (pd df): Dataframe containing columns for dimensionality reduction and category
+                    columns (list): List of column names that are included for dimensionality reduction
+                    category (str): Name of column in dataframe containing categorical Population or Patient data
+                    color_map (str): Name of MatPlotLib ColorMap used, default is Spectral
+                    algorithm (str): Name of dimensionality reduction algorithm, default is UMAP
+                    save_dir (str): Directory to save plots, default is None"""
+    cell_data = cell_data.dropna()
+
+    assert(algorithm == "UMAP" or algorithm == "PCA" or algorithm == "tSNE")
+
+    if algorithm == "UMAP":
+        reducer = umap.UMAP()
+
+        column_data = cell_data[columns].values
+        scaled_column_data = StandardScaler().fit_transform(column_data)
+        embedding = reducer.fit_transform(scaled_column_data)
+
+        fig1 = plt.figure(1)
+        plt.scatter(embedding[:, 0], embedding[:, 1], cmap=color_map,
+                    c=sns.color_palette(color_map, n_colors=len(cell_data)))
+        plt.gca().set_aspect('equal', 'datalim')
+        plt.title('UMAP projection of data', fontsize=24)
+        fig1.show()
+        if save_dir is not None:
+            plt.savefig(save_dir + "UmapVisualization.png")
```

Comment:
Recommend looking into `os.path.join` instead of simple string concatenation. I think in this case it would fail if the user specified a non-blank save_dir name because the forward-slash between save_dir and the file name would be missing. You could always just concatenate a `\` as well, but `os.path.join` ensures that (as well as cross-platform support).